src/iarc7_sensors/obstacle_detector.py

```diff
@@ -0,0 +1,468 @@
+        obst_odom.pose.pose.position.z = 0.0
+
+        obst_odom.pose.pose.orientation.x = orientation[0]
+        obst_odom.pose.pose.orientation.y = orientation[1]
+        obst_odom.pose.pose.orientation.z = orientation[2]
+        obst_odom.pose.pose.orientation.w = orientation[3]
+
+        obst_odom.twist.twist.linear.x = settings['obst_speed']
+
+        marker = Marker()
+        marker.header.stamp = obst_odom.header.stamp
+        marker.header.frame_id = 'map'
+        marker.ns = 'obstacles'
+
+        global marker_id
+        try:
+            marker_id += 1
+        except NameError:
+            marker_id = 0
+        marker.id = marker_id
+
+        marker.type = Marker.POINTS
+        marker.action = Marker.ADD
+        marker.pose.orientation.w = 1.0
+
+        marker.scale.x = 0.1
+        marker.scale.y = 0.1
+        marker.scale.z = 0.1
+
+        marker.color.r = 1.0
+        marker.color.g = 0.0
+        marker.color.b = 0.0
+        marker.color.a = 1.0
+
+        marker.lifetime = rospy.Duration(1.0)
+
+        point = Point()
+        point.x = obst_pos[0]
+        point.y = obst_pos[1]
+        marker.points.append(point)
+
+        marker_msg.markers.append(marker)
+
+        obst_msg = Obstacle()
+        obst_msg.header.stamp = obst_odom.header.stamp
+        obst_msg.odom = obst_odom
+        obst_msg.pipe_radius = settings['obst_radius']
+        obst_msg.pipe_height = settings['obst_height']
+
+        msg.obstacles.append(obst_msg)
+
+    obstacle_pub.publish(msg)
+
+    if marker_msg.markers:
+        marker_pub.publish(marker_msg)
+
+if __name__ == '__main__':
+    rospy.init_node('obstacle_detector')
+    rospy.Subscriber('scan', LaserScan, scan_callback)
+
+    obstacle_pub = rospy.Publisher('obstacles', ObstacleArray, queue_size=5)
+    marker_pub = rospy.Publisher('obstacle_markers', MarkerArray, queue_size=5)
+
+    tf_buffer = tf2_ros.Buffer()
+    tf_listener = tf2_ros.TransformListener(tf_buffer)
+
+    msg_queue = deque()
+
+    settings = {}
+    settings['min_x'] = rospy.get_param('~min_x')
+    settings['max_x'] = rospy.get_param('~max_x')
+    settings['min_y'] = rospy.get_param('~min_y')
+    settings['max_y'] = rospy.get_param('~max_y')
+    settings['min_z'] = rospy.get_param('~min_z')
+    settings['max_z'] = rospy.get_param('~max_z')
+    settings['drone_radius'] = rospy.get_param('~drone_radius')
+    settings['cluster_threshold'] = rospy.get_param('~cluster_threshold')
+    settings['arena_center_x'] = rospy.get_param('~arena_center_x')
+    settings['arena_center_y'] = rospy.get_param('~arena_center_y')
+    settings['obst_radius'] = rospy.get_param('~obst_radius')
+    settings['obst_height'] = rospy.get_param('~obst_height')
+    settings['obst_speed'] = rospy.get_param('~obst_speed')
+    settings['max_queue_size'] = rospy.get_param('~max_queue_size')
+
+    timeout = rospy.Duration(rospy.get_param('~timeout'))
+
+    rate = rospy.Rate(20)
```

Comment:
Can we have the rate be a loaded in param?

Reply:
Fixed